controllers/drain_controller.go

```diff
@@ -0,0 +1,492 @@
+func (dr *DrainReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// try to lock the node to this reconcile loop.
+	// if we are not able this means there is another loop already handling this node, so we just exist
+	if !dr.TryLockNode(req.Name) {
+		return ctrl.Result{}, nil
+	}
+
+	// configure logs
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("Reconciling Drain")
+	reqLogger.V(2).Info("node locked for drain controller", "nodeName", req.Name)
+
+	// we send to another function so the operator can have a defer function to release the lock
+	return dr.reconcile(ctx, req)
+}
+
+func (dr *DrainReconcile) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// remove the lock when we exist the function
+	defer dr.unlockNode(req.Name)
+
+	req.Namespace = vars.Namespace
+
+	// configure logs
+	reqLogger := log.FromContext(ctx)
+
+	// get node object
+	node := &corev1.Node{}
+	err := dr.Get(ctx, req.NamespacedName, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "node object doesn't exist", "nodeName", req.Name)
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "failed to get node object from api re-queue the request", "nodeName", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// get sriovNodeNodeState object
+	nodeNetworkState := &sriovnetworkv1.SriovNetworkNodeState{}
+	err = dr.Get(ctx, req.NamespacedName, nodeNetworkState)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "sriovNetworkNodeState object doesn't exist", "nodeName", req.Name)
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "failed to get sriovNetworkNodeState object from api re-queue the request", "nodeName", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// create the drain state annotation if it doesn't exist in the sriovNetworkNodeState object
+	nodeStateDrainAnnotationCurrent, NodeStateDrainAnnotationCurrentExist := nodeNetworkState.Annotations[constants.NodeStateDrainAnnotationCurrent]
+	if !NodeStateDrainAnnotationCurrentExist {
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeStateDrainAnnotationCurrent = constants.DrainIdle
+	}
+
+	// create the drain state annotation if it doesn't exist in the node object
+	nodeDrainAnnotation, nodeDrainAnnotationExist := node.Annotations[constants.NodeDrainAnnotation]
+	if !nodeDrainAnnotationExist {
+		err = utils.AnnotateObject(node, constants.NodeDrainAnnotation, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeDrainAnnotation = constants.DrainIdle
+	}
+
+	reqLogger.V(2).Info("Drain annotations", "nodeAnnotation", nodeDrainAnnotation, "nodeStateAnnotation", nodeStateDrainAnnotationCurrent)
+
+	// Check the node request
+	if nodeDrainAnnotation == constants.DrainIdle {
```

Comment:
All in all, it's quite confusing to see why we are tracking two node drain annotations.

Reply:
it's part of the design proposal. We need two because two different components are updating so every component should know the current status of is flow

Reply:
Can you elaborate on why we can't have one instead of two annotations?

Reply:
sure :)

so example: 
1. daemon move the annotation to reqDrain
2. then the operator moves to draining the annotation
3. user remove the policy
4. daemon moves the annotation to idle

you are stuck in a middle of a drain because the operator doesn't know where he was

in the case we have 2 annotations
1. daemon move the annotation to reqDrain
2. then the operator mark  draining is own annotation
3. user remove the policy
4. daemon moves the annotation to idle
5. operator see the change and run the un-drain 

let me know if that is clear.